script/git-ship

```diff
@@ -1,26 +1,36 @@
 #!/usr/bin/env perl
 use Applify;
 use App::git::ship;
+use Mojo::Util qw{monkey_patch};
 
 documentation 'App::git::ship';
 version 'App::git::ship';
 
-app {
-  my $self   = shift;
-  my $action = shift || 'ship';
-  my $ship   = App::git::ship->new;
+subcommand build => 'build the project - runs tests' => _applify('build');
+subcommand clean => 'scrub the decks' => _applify('clean');
+subcommand coverage => 'test coverage' => _applify('test_coverage');
+subcommand start => 'start a project' => _applify('start');
+subcommand upload => 'set sail' => _applify('ship');
 
-  $action =~ s/-/_/g;
+sub _applify {
+  my ($pkg, $export, $class) = ('App::git::ship', shift, caller);
+  my ($sc, @args) = @ARGV;
+  return sub {
+    my $applify = shift;
+    my $detected = $pkg->new->detect($sc eq 'start' ? grep { m{/} } @args : ());
+    $applify->documentation( $detected );
+    $applify->extends( $detected );
 
-  if ($action eq 'start') {
-    $ship->start(@_);
-  }
-  else {
-    $ship = $ship->detect->new;
-    $ship->config;
-    $ship->abort("Unknown action: $action") if $action !~ /^[a-z_]+$/ or !$ship->can($action);
-    $ship->$action(@_);
-  }
+    my $entry_point = join '_', 'command', $applify->subcommand;
+    monkey_patch $class, $entry_point
+      => sub { shift->$export(@_); } if $detected->can($export);
+  };
```

Comment:
I do find this very hard to read. Why not do something like this instead: (Not tested)

```
subcommand build    => 'Build the project'        => \&detect_before_run;
subcommand clean    => 'Clean out unwanted files' => \&detect_before_run;
subcommand coverage => 'Test coverage'            => \&detect_before_run;
subcommand start    => 'Start a project'          => sub { };

sub detect_before_run { 
  shift->documentation(App::git::ship->new->detect);
}

sub command_start { 
  my $self = shift;
  App::git::ship->new->start(@_);
  return 0;
} 

for my $method (qw(build clean coverage upload)) {
  monkey_patch __PACKAGE__, "command_$method" => sub {
    shift->_script->documentation->new->$method(@_);
    return 0;
  }; 
}
```

Reply:
I don't even think we need money_patch:

```
sub command_build    { shift->_script->documentation->new->build(@_);    0 }
sub command_clean    { shift->_script->documentation->new->clean(@_);    0 }
sub command_coverage { shift->_script->documentation->new->coverage(@_); 0 }
sub command_upload   { shift->_script->documentation->new->upload(@_);   0 }
```